### PR TITLE
Integrate Clang-Tidy Workflow for DD Package

### PR DESCRIPTION
**Pull request by hillmich**
This PR introduces a new CI job to check for compliance via Clang Tidy.

This includes checking variable and function names, warnings for C-style casts and a few more.

(Recreated from #103)

### CODE REVIEW

File: .clang-tidy
Comment (burgholzer) [inline .clang-tidy:26]: See why this is excluded. However, reading https://releases.llvm.org/13.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/performance-no-int-to-ptr.html, I was wondering whether carefully looking at the relevant places there is some room for improvement. As I interpret it from the link above, there might be ways to give the compiler more transparent information and potential for optimization.

File: .github/workflows/cpp-linter.yml
Comment (burgholzer) [inline .github/workflows/cpp-linter.yml:36]: don't forget to change this back before merging
```suggestion
          --files-changed-only=true \
```

File: include/dd/ComplexValue.hpp
Comment (burgholzer) [inline include/dd/ComplexValue.hpp:73]: Just a nitpick. This might be cleaner as
```suggestion
        static auto getLowestFraction(const double x, const std::uint64_t maxDenominator = 1U << 10, const fp tol = dd::ComplexTable<>::tolerance()) {
```
Or does this raise an implicit conversation warning again?

File: include/dd/Control.hpp
Comment (burgholzer) [inline include/dd/Control.hpp:16]: Would it be too inconvenient to change these? I would assume they are not used in that many places.

File: include/dd/Edge.hpp
Comment (burgholzer) [inline include/dd/Edge.hpp:36]: does the const really work here? I believe I had troubles with that in the past in our top-level repos.

File: include/dd/Node.hpp
Comment (burgholzer) [inline include/dd/Node.hpp:36]: Try to stay consistent with `NOLINTNEXTLINE` and `NOLINT`. Linting on the previous struct was ignored differently.

**Review by burgholzer — CHANGES_REQUESTED** [review]
Thanks for all the changes! It's great to see consistency across the repository.
I got a couple of minor comments/requests.

Yet to make it through the Package.hpp file. Expect another review of that file soon.

File: .clang-tidy
Comment (hillmich) [reply .clang-tidy:26]: I did some research on this and would stay with the current implementation handling the tagged pointers. I think we would have to cast to `char*` anyway to change the LSB with pointer arithmetic and LLVM uses the same scheme as we are using in [Error.h](https://llvm.org/doxygen/Support_2Error_8h_source.html) (CTRL+F `void setChecked(bool V)`).

File: include/dd/Control.hpp
Comment (hillmich) [reply include/dd/Control.hpp:16]: It's more about usage outside of the package. As we expose this towards the public I'm slightly in favor of keeping the naming as is.

File: .clang-tidy
Comment (burgholzer) [reply .clang-tidy:26]: Fully trust your judgement on this and agree that it seems reasonable to keep it that way. 👍🏻 feel free to resolve this.

File: include/dd/Control.hpp
Comment (burgholzer) [reply include/dd/Control.hpp:16]: True. Would certainly be a breaking change. The question is: how many people are affected by that change and how hard is it to adapt. I would argue that changing capitalization of a tiny struct's members wouldn't be too much of a hassle. In the end, I'll just leave it to you to decide. I am fine with either solution.

File: include/dd/Package.hpp
Comment (burgholzer) [inline include/dd/Package.hpp:2517]: Why not just use `std::uint16_t` instead of ignoring the Linter warning? Or `std::uint_fast16_t` if you are concerned about the former not being standard mandated (although we are using the fixed width types elsewhere in our projects and, hence, this wouldn't be the first place they are introduced).

**Review by burgholzer — CHANGES_REQUESTED** [review]
Just finished going through the leftover file. Only two small questions/comments there. Overall there should not be much to address here 👍🏻

File: .clang-tidy
Comment (hillmich) [reply .clang-tidy:26]: I resolve as discussed

File: include/dd/Package.hpp
Comment (hillmich) [reply include/dd/Package.hpp:2517]: Good catch. For consistency with the types I'll change `short` to `std::size_t`.

File: include/dd/Control.hpp
Comment (hillmich) [reply include/dd/Control.hpp:16]: As discussed, we'll keep it as proposed. Future plan will probably necessitate changes here anyway and we don't need to break things twice.

File: include/dd/Package.hpp
Comment (hillmich) [reply include/dd/Package.hpp:108]: The constants will likely stay as proposed, so we'll keep the change :)

File: include/dd/Edge.hpp
Comment (hillmich) [reply include/dd/Edge.hpp:36]: DDSIM builds and tests are successful, so we'll keep it const :)

**Review by burgholzer — APPROVED** [review]
LGTM 👍🏼 Thanks for pushing this!

**Pull request merged**
